enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -176,51 +178,59 @@ Are the same in that context, the container has access to the same number of sha
-NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
-Additionally, since the plugin is part of OCP infrastructure pods, NTO annotates the pods with:
-```yaml
-target.workload.openshift.io/management: {
-  "effect":"PreferredDuringScheduling"
+The reason for specifying a device, and not an annotation,
+is to allow control on the number of workloads that can request for shared cpus.
+By populating shared cpu devices as the number of maximum workloads.
+If the number of shared cpu devices got exhausted, any new workload that requests for shared cpus device will move
+to pending state by the scheduler.
+
+Once workload request completed successfully it gets access to shared cpus, and two new environment variables
+will be present under the container's environment variables:
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` = specifies the core ids of the shared cpus
+`OPENSHIFT_ISOLATED_CPUS=<CPU-IDs>` = specifies the core ids of the isolated cpus
+Those environment variables help the application's user to pin its processes/threads to the desired CPU set. 
+
+#### Kubelet
+Kubelet needs to be updated to advertise the `openshift.io/shared-cpus` devices 
+as [Extended Resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources).
+Kubelet will look for a configuration file `/etc/kubernetes/openshift-shared-cpus`, to enable the device advertisement.
+The configuration file should be look like this (the number of devices might be varied)
+``` json
+{
+  "shared-cpus": {
+     "quantity": "15"
+  }
 }
 ```
-See https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md for details.
 
-Since both the components are related,
-mixed-cpu-node-plugin [source code](https://github.com/openshift-kni/mixed-cpu-node-plugin) would be vendored under NTO.
-By having different main.go + container entrypoint, NTO image can be also used as the
-mixed-cpu-node-plugin image. 
-This way saves the overhead of having a separate container image with all the maintenance cost that comes with it.
+#### API Server Admission Hook
+A new admission hook for [OpenShift Kubernetes API Server](https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission)
+will be added for handling the following:
+1. In case a user specifies more than a single device, it mutates the number to be one (since we treat it as a boolean value anyway).
+2. It adds an annotation `cpu-shared.crio.io` that will be used to tell the runtime that a shared cpus were requested. 
+The `cpu-shared.crio.io` annotation needs to be added under the performance-runtime [allowed_annotation](https://github.com/openshift/cluster-node-tuning-operator/blob/master/assets/performanceprofile/configs/99-runtimes.conf#L20)   
+
+#### CRI-O
+CRI-O should update with a new performance hook to support the shared-cpu logic.
```

Comment:
Can you add a little detail about the cri-o logic. How will cri-o handle cases where there are multiple containers in the same pod? If the cpu-shared.crio.io annotation is at the pod level, but each container requests the new shared-cpus resource at the container level, what happens when some containers request shared-cpus but other containers in the same pod don't?

Reply:
> If the cpu-shared.crio.io annotation is at the pod level, but each container requests the new shared-cpus resource at the container level

The user should request for shared cpu at the container level:
```
resources:
  request: 
    shared-cpu/openshift.io
```
The `cpu-shared.crio.io` annotation gets created automatically by the mutation web hook when it sees the shared-cpu request under the container.
we can extend the annotation to include the container name so CRI-O would know exactly to which container it should add the cpus.

Reply:
> we can extend the annotation to include the container name so CRI-O would know exactly to which container it should add the cpus

The problem is that the container name is always changing and I'm not sure we can add an annotation with wildcard under the `allowed_annotation`: https://github.com/openshift/cluster-node-tuning-operator/blob/master/assets/performanceprofile/configs/99-runtimes.conf#L20

Reply:
on the cri-o side, that will work. We currently have a similar mechanism for the `io.containers.trace-syscall` annotation. To match a allowed annotation, a prefix match is done. Then each allowed annotation defines it's own behavior on how to interpret the annotation